core/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.apache.lucene.search.ReferenceManager;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.translog.Translog;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.LinkedTransferQueue;
+import java.util.function.Consumer;
+import java.util.function.IntSupplier;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Allows for the registration of listeners that are called when a change becomes visible for search. This functionality is exposed from
+ * {@link IndexShard} but kept here so it can be tested without standing up the entire thing. 
+ */
+class RefreshListeners {
+    /**
+     * Refresh listeners. While they are not stored in sorted order they are processed as though they are.
+     */
+    private final LinkedTransferQueue<Tuple<Translog.Location, Consumer<Boolean>>> refreshListeners = new LinkedTransferQueue<>();
+
+    private final IntSupplier getMaxRefreshListeners;
+    private final Runnable forceRefresh;
+    private final Consumer<Runnable> fireListener;
```

Comment:
it's unclear that this consumer executes concurrently - I think we should make ti more explicit and just use `Executor` it has also one method and be be easily mocked?